authenticate/handlers.go

```diff
@@ -187,19 +194,33 @@ func (a *Authenticate) getOAuthCallback(w http.ResponseWriter, r *http.Request)
 	if err != nil {
 		return nil, httputil.Error("malformed state", http.StatusBadRequest, err)
 	}
-	// split state into its it's components (nonce:redirect_uri)
-	statePayload := strings.SplitN(string(bytes), ":", 2)
-	if len(statePayload) != 2 {
-		return nil, fmt.Errorf("state malformed, size: %d", len(statePayload))
+
+	// split state into its it's components, e.g.
+	// (nonce|timestamp|redirect_url|encrypted_data(redirect_url)+mac(nonce,ts))
+	statePayload := strings.SplitN(string(bytes), "|", 3)
```

Comment:
There's a small dark part of me that thinks we should maybe use an off the shelf ser/deser...this should be sufficient, but it feels a little fiddly.

Reply:
I considered it as well but in this case I want/need to be able to pluck out the associated cleartext data first (the csrf and timestamp) before decrypting/verifying the mac. The [CSRF nonce is checked in middleware](https://github.com/pomerium/csrf/blob/f3318380bad3b451772163283269d0454da833b3/helpers.go#L121).   So the process is:

1. Middleware checks for the cleartext CSRF in the state query param. If invalid, reject.
1.  Check cleartext timestamp within leeway, if invalid, reject. (mac not checked yet!)
1.  Decrypt the payload (redirect_uri) and verify the authenticity of the CSRF/timestamp. If any fail, reject. 

The only weird edge-case I could see is the IdP intentionally forging the CSRF or timestamp, but that would just result in rejection ... or maybe a denial of service to the tend user who now can't log in. Of course, if you don't trust your identity provider, you've got bigger issues here. 

Hopefully that makes sense?

Reply:
Yep makes sense.  I think in practice this is fine.